nordic nrf-timer: Expose max frequency as DT property and use it in nrf counter driver

aescolar opened the pull request:
The counter driver needs to know what is the maximum counting frequency of each timer peripheral.
Let's add it to its DT, and let's use it in the driver directly instead of relaying on the HAL working it out.

aescolar committed:
dts: nordic nrf-timer: Expose max frequency as DT property

The counter driver needs to know what is the maximum
counting frequency of each timer peripheral.
Let's add it to its DT.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers: counter_nrfx_timer.c: Get frequency directly from DT

Instead of working it from a macro from the HAL
based on the DT address, let's just get the maximum
frequency directly from DT.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

masz-nordic commented:
Can you add checks that validate values in DT against `NRF_TIMER_BASE_FREQUENCY_GET(p_reg)`?
Also, nit - missing `:` in the second commit.

masz-nordic commented:
Right, but can you do something similar to:
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/validate_base_addresses.c#L127
that checks if both values are identical - we need to make sure that stuff in DT is identical to HW configuration.
It can be even disabled for simulation targets if that's a problem.

decsny commented on dts/bindings/timer/nordic,nrf-timer.yaml:
```diff
@@ -24,10 +24,15 @@ properties:
 
+  max-frequency:
+    type: int
+    default: 16
```
default value needs justification in the description

aescolar replied:
Thanks @decsny . Done

aescolar commented:
> Right, but can you do something similar to..

@masz-nordic I don't think there is a direct way of doing that today.

Note that `NRF_TIMER_BASE_FREQUENCY_GET()`, unfortunately, does not rely on the MDK 
https://github.com/nrfconnect/nrfx/blob/master/hal/nrf_timer.h#L178-L220
Instead the nrfx hal keeps in macros the knowledge of which timer runs at which frequency based on the timer name.
During a discussion with @nika-nordic it seemed that macro implementation would be better changed.

To be able to rely on the MDK provided info, effectively we would need to implement a new `NRF_TIMER_BASE_FREQUENCY_GET` which would use the MDK `TIMER*_PCLK_MHZ` macros. (also note those are only provided for nRF54 devices, so we'd need that macro to fallback to 16MHz when that is not provided).
If that is desired, it seems to me it should be done as a separate task in the nrfx.

anangl commented on dts/bindings/timer/nordic,nrf-timer.yaml:
```diff
@@ -24,10 +24,20 @@ properties:
+    type: int
+    default: 16
+    description: |
+        Maximum timer frequency in MHz.
```
What's the benefit of having it in MHz? Couldn't we use just Hz, like for example in SPI: https://github.com/zephyrproject-rtos/zephyr/blob/b116bc7da03ce6a05ce189a7bc36a45c397d5fff/dts/bindings/spi/nordic%2Cnrf-spi-common.yaml#L18-L23 or for `clock-frequency` in several bindings?

aescolar replied:
Changed to Hz

aescolar committed:
drivers: counter_nrfx_timer.c: Check max-frequency DT property

Check that the max-frequency DT property matches the
one provided by the nrfx HAL.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar commented:
Added an extra commit as per @masz-nordic [request](https://github.com/zephyrproject-rtos/zephyr/pull/72381#issuecomment-2096334198)

anangl commented on drivers/counter/counter_nrfx_timer.c:
```diff
@@ -417,6 +417,14 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
 
+#if !defined(CONFIG_SOC_SERIES_BSIM_NRFXX)
+#define CHECK_MAX_FREQ(idx)									\
+		BUILD_ASSERT(DT_INST_PROP(idx, max_frequency) ==				\
+			NRF_TIMER_BASE_FREQUENCY_GET((NRF_TIMER_Type *)DT_INST_REG_ADDR(idx)))
```
The indentation here could be one tab less.
